allennlp/data/dataset_readers/huggingface_datasets_reader.py

```diff
@@ -0,0 +1,190 @@
+from typing import Iterable, Optional
+
+from allennlp.data import DatasetReader, Token, Field
+from allennlp.data.fields import TextField, LabelField, ListField
+from allennlp.data.instance import Instance
+from datasets import load_dataset, Dataset
+from datasets.features import ClassLabel, Sequence, Translation, TranslationVariableLanguages
+from datasets.features import Value
+
+
+class HuggingfaceDatasetSplitReader(DatasetReader):
+    """
+    This reader implementation wraps the huggingface datasets package
+    to utilize it's dataset management functionality and load the information in AllenNLP friendly formats
+    Note: Reader works w.r.t to only one split of the dataset,
+    i.e. you would need to create separate reader for separate splits
+
+    Following dataset and configurations have been verified and work with this reader
+
+            Dataset                       Dataset Configuration
+            `xnli`                        `ar`
+            `xnli`                        `en`
+            `xnli`                        `de`
+            `xnli`                        `all_languages`
+            `glue`                        `cola`
+            `glue`                        `mrpc`
+            `glue`                        `sst2`
+            `glue`                        `qqp`
+            `glue`                        `mnli`
+            `glue`                        `mnli_matched`
+            `universal_dependencies`      `en_lines`
+            `universal_dependencies`      `ko_kaist`
+            `universal_dependencies`      `af_afribooms`
+            `afrikaans_ner_corpus`        `NA`
+            `swahili`                     `NA`
+            `conll2003`                   `NA`
+            `dbpedia_14`                  `NA`
+            `trec`                        `NA`
+            `emotion`                     `NA`
+    """
+
+    def __init__(
+        self,
+        max_instances: Optional[int] = None,
+        manual_distributed_sharding: bool = False,
+        manual_multiprocess_sharding: bool = False,
+        serialization_dir: Optional[str] = None,
+        dataset_name: str = None,
+        split: str = "train",
```

Comment:
I think it makes the most sense if we use the `file_path` parameter from the reader for `split`. We do this in many other cases too. `file_path` is sort of a legacy name for what that parameter does.

If you do this, you can specify the reader like normal, and then for the training set and the validation set you just put in the name of the split.

Reply:
It means you have to delay the `load_dataset` call, but that's not a problem, maybe even preferred?

Reply:
Makes sense, I can use the `file_path` to read from the specific split indicated.
Let me think this through, I am little rusty on py generator stuff, based on file_path it would need to return from a different split in a generator style. 

Coming to delay of `load_dataset`, we can go in two ways
1. Load all splits during `__init__` it would mean the first `_read` for split will not be slow
2. Load Split on the first `_read` in which case that one read will be really really slow. 

I would prefer 1 over 2, since it is clear but, once Reader object is ready all splits are ready for quick reads. 
Not sure about how _datasets_ handles the memory load, can dig deeper sometime later if required.

Reply:
If the HF side always downloads all splits, then I can see why you might want to read it in `__init__`. But I suspect it does not. There is no point in downloading a huge training set if you just want to do evaluation. That's why I prefer doing it in `_read()`.

Reply:
We have the control to download split or all splits, just that it returns `Dataset` (one split) or `DatasetDict` (multiple splits) based on the choice. Changing so a split is loaded through hf in reader, introduced a `pre_load ` flag that can be used if required to download all splits in `_init_` if so desired. I can remove that you believe it won't be useful.

Reply:
We have the control to download split or all splits, just that it returns `Dataset` (one split) or `DatasetDict` (multiple splits) based on the choice. Changed so a split is loaded through hf in `_read`, introduced a `pre_load ` flag that can be used if required to download all splits in `_init_` if so desired. I can remove that you believe it won't be useful.